crates/oxc_linter/src/rules/react_hooks/exhaustive_deps.rs

```diff
@@ -0,0 +1,3454 @@
+                }
+
+                let cur_skip_reporting_dependency = self.skip_reporting_dependency;
+                self.skip_reporting_dependency = true;
+                self.visit_expression(&it.object);
+                self.skip_reporting_dependency = cur_skip_reporting_dependency;
+            }
+            // this means that some part of the chain could not be analyzed
+            // for example `foo.bar.baz().abc`. `baz()` cannot be statically analyzed
+            // instead, continue to go down, looking at the object to gather dependencies
+            Err(()) => {
+                self.visit_expression(&it.object);
+            }
+        }
+    }
+
+    fn visit_identifier_reference(&mut self, ident: &IdentifierReference<'a>) {
+        if self.skip_reporting_dependency {
+            return;
+        }
+
+        // TODO: is there a better way?
+        let ident = unsafe {
+            std::mem::transmute::<&IdentifierReference<'_>, &'a IdentifierReference<'a>>(ident)
+        };
+        self.found_dependencies.insert(Dependency {
+            iref: ident,
+            chain: vec![],
+            symbol_id: self.semantic.symbols().get_reference(ident.reference_id()).symbol_id(),
+        });
+
+        if let Some(decl) = get_declaration_of_variable(ident, self.semantic) {
+            let is_set_state_call = match decl.kind() {
+                AstKind::VariableDeclarator(var_decl) => {
+                    if let Some(Expression::CallExpression(call_expr)) = &var_decl.init {
+                        if let Some(name) = func_call_without_react_namespace(call_expr) {
+                            name == "useState" || name == "useReducer"
+                        } else {
+                            false
+                        }
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            };
+
+            if is_set_state_call
+                && self.stack.iter().all(|kind| {
+                    !matches!(kind, AstKind::Function(_) | AstKind::ArrowFunctionExpression(_))
+                })
+            {
+                self.set_state_calls.push(ident);
+            }
+        }
+    }
+}
+
+fn is_inside_effect_cleanup(stack: &[AstKind<'_>]) -> bool {
+    let mut iter = stack.iter().rev();
+    let mut cur: Option<&AstKind<'_>> = iter.next();
+    loop {
+        match cur {
+            Some(AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)) => {
+                break;
+            }
+            Some(_) => cur = iter.next(),
+            None => {
+                return false;
+            }
+        }
+    }
+    cur = iter.next();
+    loop {
+        match cur {
+            Some(AstKind::ReturnStatement(_)) => {
+                return true;
+            }
+            Some(_) => cur = iter.next(),
+            None => {
+                return false;
+            }
+        }
+    }
+}
+
+#[test]
+fn test() {
```

Comment:
there are enough tests here that it might be a good idea to move it into a separate file? not sure what our standards here though are @Boshen

Reply:
1 file is fine, makes it easy to search for things.